File: frontend/lib/outreach-contacts.ts
// API client for the Outreach Contacts page (/outreach/contacts).
// Mirrors backend/app/schemas/outreach_contacts.py.

import { apiRequest } from "@/lib/api";

export type OutreachEntityKind =
  | "broker_dealer"
  | "advisor"
  | "institutional_investor";

export interface OutreachContactsFirmRow {
  entity_kind: OutreachEntityKind;
  entity_id: number;
  name: string;
  contact_count: number;
  with_email_count: number;
  with_phone_count: number;
  // Count of Email-Extractor discovered_email rows linked to this firm. A
  // parallel source to the typed contact triad above (kept separate, never
  // folded in) -- powers the "N extracted" summary pill and the lazy-loaded
  // "Extracted emails" sub-section on expand. Firms now appear in the list
  // when contact_count > 0 OR discovered_email_count > 0.
  discovered_email_count: number;
  last_enriched_at: string | null;
  last_gap_fill_attempt_at: string | null;
  gap_fill_in_progress: boolean;
}

export interface OutreachContactsFirmsResponse {
  items: OutreachContactsFirmRow[];
  total: number;
}

export interface OutreachContactPerson {
  contact_id: number;
  name: string;
  title: string | null;
  email: string | null;
  phone: string | null;
  linkedin_url: string | null;
  enriched_at: string | null;
}

export interface OutreachContactsFirmDetailResponse {
  entity_kind: OutreachEntityKind;
  entity_id: number;
  entity_name: string;
  items: OutreachContactPerson[];
}

// An Email-Extractor discovered_email row scoped to a firm. Surfaced beneath
// the typed contacts in a separate "Extracted emails" sub-section. The
// /discovered-emails endpoint returns a bare array (no envelope), unlike the
// /persons response above.
export interface DiscoveredContactRow {
  id: number;
  email: string;
  enriched_name: string | null;
  enriched_title: string | null;
  enriched_phone: string | null;
  enriched_linkedin_url: string | null;
  enrichment_status: string;
  source: string | null;
  confidence: number | null;
  created_at: string;
}

export interface GapFillFirmResponse {
  run_id: number | null;
  status: string;
  entity_kind: OutreachEntityKind;
  entity_id: number;
  reason?: string | null;
}

// Discriminates the origin of an email-search row: a typed contact row
// (ExecutiveContact / AdvisorContact / InvestorContact) vs. an
// Email-Extractor discovered_email row. Mirrors the backend EmailSource.
export type OutreachEmailSource = "contact" | "extracted";

// One flat email row returned by the contacts page "email mode" search
// (GET /outreach/contacts/email-search). Mirrors
// backend/app/schemas/outreach_contacts.py::EmailSearchResult. Unlike the
// firm-card path, this is a per-email row attributed to its owning firm via
// entity_kind + entity_id + firm_name. `phone` is only set on a typed
// contact (source="contact"); `contact_id` is set on the same.
export interface EmailSearchResult {
  entity_kind: OutreachEntityKind;
  entity_id: number;
  firm_name: string;
  owner_name: string | null;
  title: string | null;
  email: string;
  phone: string | null;
  source: OutreachEmailSource;
  contact_id: number | null;
}

export interface EmailSearchResponse {
  results: EmailSearchResult[];
  total: number;
  page: number;
  limit: number;
  has_more: boolean;
}

export interface SearchOutreachContactsByEmailParams {
  q: string;
  page?: number;
  limit?: number;
}

export interface ListOutreachContactsFirmsParams {
  entity_kind?: OutreachEntityKind;
  q?: string;
  page?: number;
  limit?: number;
}

export async function listOutreachContactsFirms(
  params: ListOutreachContactsFirmsParams = {},
): Promise<OutreachContactsFirmsResponse> {
  const search = new URLSearchParams();
  if (params.entity_kind) search.set("entity_kind", params.entity_kind);
  if (params.q) search.set("q", params.q);
  if (params.page) search.set("page", String(params.page));
  if (params.limit) search.set("limit", String(params.limit));
  const qs = search.toString();
  const url = qs
    ? `/api/v1/outreach/contacts/firms?${qs}`
    : "/api/v1/outreach/contacts/firms";
  return apiRequest<OutreachContactsFirmsResponse>(url);
}

export async function listOutreachContactsFirmPersons(
  kind: OutreachEntityKind,
  id: number,
): Promise<OutreachContactsFirmDetailResponse> {
  return apiRequest<OutreachContactsFirmDetailResponse>(
    `/api/v1/outreach/contacts/firms/${kind}/${id}/persons`,
  );
}

// Pass the SAME `kind` the firm row carries (the persons wrapper above uses
// it too). Returns a bare array of discovered emails for the firm.
export async function listOutreachContactsFirmDiscoveredEmails(
  kind: OutreachEntityKind,
  id: number,
): Promise<DiscoveredContactRow[]> {
  return apiRequest<DiscoveredContactRow[]>(
    `/api/v1/outreach/contacts/firms/${kind}/${id}/discovered-emails`,
  );
}

export async function gapFillFirmContacts(
  kind: OutreachEntityKind,
  id: number,
): Promise<GapFillFirmResponse> {
  return apiRequest<GapFillFirmResponse>(
    `/api/v1/outreach/contacts/firms/${kind}/${id}/gap-fill`,
    { method: "POST" },
  );
}

// "Email mode" search for the contacts page: when the user's query is an
// email address, return the matching EMAIL rows directly instead of firm
// cards. Same fetch wrapper / credentials / error handling as the sibling
// fns above. A blank/whitespace q short-circuits to an empty page on the BE
// (no 422), but callers should only invoke this in email mode anyway.
export async function searchOutreachContactsByEmail(
  params: SearchOutreachContactsByEmailParams,
): Promise<EmailSearchResponse> {
  const search = new URLSearchParams();
  search.set("q", params.q);
  if (params.page) search.set("page", String(params.page));
  if (params.limit) search.set("limit", String(params.limit));
  return apiRequest<EmailSearchResponse>(
    `/api/v1/outreach/contacts/email-search?${search.toString()}`,
  );
}

File: backend/app/schemas/outreach_contacts.py
"""Schemas for the Outreach Contacts page (`/outreach/contacts`).

This is a separate namespace from ``schemas/vault.py``'s outreach
schemas because the new page returns a richer person shape (phone,
LinkedIn, enriched_at) plus list-level summary counts that the existing
send-flow autocomplete doesn't need. Kept in its own module so widening
fields on this page can't accidentally leak into the send flow.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


EntityKind = Literal["broker_dealer", "advisor", "institutional_investor"]


class OutreachContactsFirmRow(BaseModel):
    """One firm row in the persons-by-firm list. One row covers any of
    the three entity kinds via ``entity_kind`` + ``entity_id``."""

    entity_kind: EntityKind
    entity_id: int
    name: str
    contact_count: int
    with_email_count: int
    with_phone_count: int
    # Count of Email-Extractor ``discovered_email`` rows linked to this
    # firm. A PARALLEL source from the typed contact tables -- the three
    # counts above stay typed-only; this one is kept separate so the FE
    # can render a distinct "extracted emails" pill without double-
    # counting. For institutional investors this counts through the
    # resolved ``advisor_id`` (0 when the investor has no advisor link,
    # since ``discovered_email`` carries no ``investor_id``).
    discovered_email_count: int
    last_enriched_at: datetime | None
    last_gap_fill_attempt_at: datetime | None
    # True when a gap-fill PipelineRun for this firm is currently
    # ``queued`` or ``running``. Used by the FE to disable the row's
    # "Enrich all" button so the user doesn't double-fire while a job
    # is in flight.
    gap_fill_in_progress: bool


class OutreachContactsFirmsResponse(BaseModel):
    items: list[OutreachContactsFirmRow]
    total: int


class OutreachContactPerson(BaseModel):
    """One contact row inside an expanded firm. Widens the shape used
    by ``/firms/contacts`` (which is email-only) so the persons-by-firm
    accordion can render phone-only contacts and surface enriched_at."""

    contact_id: int
    name: str
    title: str | None
    email: str | None
    phone: str | None
    linkedin_url: str | None
    enriched_at: datetime | None


class OutreachContactsFirmDetailResponse(BaseModel):
    entity_kind: EntityKind
    entity_id: int
    entity_name: str
    items: list[OutreachContactPerson]


class OutreachDiscoveredEmailRow(BaseModel):
    """One Email-Extractor ``discovered_email`` row surfaced under a firm.

    A parallel source from the typed contact rows in
    :class:`OutreachContactPerson`: these come straight from the Email
    Extractor's scans (``discovered_email`` table) and carry no typed
    ``contact_id`` -- the FE wires outreach off the email address via the
    adhoc compose path. ``email`` is the discovered address that keyed the
    scan; the ``enriched_*`` fields are whatever provider enrichment
    attached afterwards (may be NULL on an unenriched row)."""

    id: int
    email: str
    enriched_name: str | None
    enriched_title: str | None
    enriched_phone: str | None
    enriched_linkedin_url: str | None
    enrichment_status: str
    source: str | None
    confidence: float | None
    created_at: datetime


class GapFillFirmResponse(BaseModel):
    """202 ACCEPTED response from the unified gap-fill dispatch endpoint."""

    run_id: int | None
    status: str
    entity_kind: EntityKind
    entity_id: int
    reason: str | None = None


# Distinct ``source`` discriminator for an email row in the email-search
# results: a typed contact row (ExecutiveContact / AdvisorContact /
# InvestorContact -- "Generate More Details" enrich output) vs. an
# Email-Extractor ``discovered_email`` row.
EmailSource = Literal["contact", "extracted"]


class EmailSearchResult(BaseModel):
    """One email/contact row returned by ``/outreach/contacts/email-search``.

    A FLAT, per-email row (not a firm row) that the FE renders when the
    user searches the contacts page by an email address. Unifies the two
    sources behind one shape: a typed contact (``source="contact"``,
    carries ``contact_id`` + any typed ``phone``) and an Email-Extractor
    ``discovered_email`` (``source="extracted"``, no ``contact_id``;
    ``phone`` is always None since the discovered shape is keyed off the
    address, not a typed channel record).

    ``entity_kind`` + ``entity_id`` + ``firm_name`` identify the owning
    firm so the FE can link the row back to its firm. ``owner_name`` /
    ``title`` come from the contact (``name`` / ``title``) or the
    discovered row (``enriched_name`` / ``enriched_title``)."""

    entity_kind: EntityKind
    entity_id: int
    firm_name: str
    owner_name: str | None
    title: str | None
    email: str
    phone: str | None
    source: EmailSource
    contact_id: int | None


class EmailSearchResponse(BaseModel):
    """Paginated wrapper for the email-search results.

    Mirrors the firms-list pagination contract: ``total`` is the full
    match count across both sources (after dedup), ``page`` / ``limit``
    echo the request, and ``has_more`` tells the FE whether another page
    exists without it recomputing from ``total``."""

    results: list[EmailSearchResult]
    total: int
    page: int
    limit: int
    has_more: bool
